### ISSUE TITLE
Possible bug: prob_ratio always 1

### ISSUE THREAD

**Issue by magnusja** [reinforceio/tensorforce#312]
Hey guys,

I am not a 100% sure if it is a bug but I find it certainly a little odd. I added a tf.Print to see the probability ratio between old and new policy in PG methods and this is always exactly 1 for me. This is the case for master and the memory branch. I feel like for the very first time of an update procedure this makes sense, but then after going through several mini batches the two distributions should certainly a little different?
I think it is necessary to cache the parameters of the old (behavioral/sampling) distribution in order to perform a full forward pass with the current mini batch and get the correct distribution of the old policy. This is probably not a major issue because usually, this ratio should be very close to 1.

```python
def tf_pg_loss_per_instance(self, states, internals, actions, terminal, reward, update):
        embedding = self.network.apply(x=states, internals=internals, update=update)
        prob_ratios = list()
        for name, distribution in self.distributions.items():
            distr_params = distribution.parameterize(x=embedding)
            log_prob = distribution.log_probability(distr_params=distr_params, action=actions[name])
            # works the same?
            # fixed_distr_params = tuple(tf.stop_gradient(input=x) for x in distr_params)
            # fixed_log_prob = distribution.log_probability(distr_params=fixed_distr_params, action=actions[name])
            fixed_log_prob = tf.stop_gradient(input=log_prob)
            prob_ratio = tf.exp(x=(log_prob - fixed_log_prob))
            prob_ratio = tf.Print(prob_ratio, [tf.reduce_mean(prob_ratio)], 'prob_ratio')
            collapsed_size = util.prod(util.shape(prob_ratio)[1:])
            prob_ratio = tf.reshape(tensor=prob_ratio, shape=(-1, collapsed_size))
            prob_ratios.append(prob_ratio)
```

EDIT:
Ok after a little of thinkking, it seems that we only need to store the distribution parameters somwhere in between batches?

**Comment by Mazecreator** [comment]
I think @AlexKuhnle documented this on gitter, you might be able to find more details there, but this will give 1.0 as the answer.  What it is designed to do is to create a gradient to update weights.  This is due to the tf.stop_gradient() statement which blocks the gradient on one item but exists on the other.  So this only shows during the optimization phase and not the forward pass as you will get the 1.0 you see with your test.

**Comment by magnusja** [comment]
Oh wow, yeah I found it, thanks! I think this is certainly worth a comment in the code.

**Comment by Mazecreator** [comment]
Ya, using kind of a side affect for something not on the forward pass.

**Comment by magnusja** [comment]
Okay, one follow up question still, when I am using PPO and minibatches I have very minibatch collected from the same behavioral policy. Would this ratio bee wrong in between two minibatches because `log_prob` is already being updated with the previous minibatch, but in my memory, there are still rollouts from the initial policy?

**Comment by AlexKuhnle** [comment]
As @Mazecreator says, it is confusing that it is always 1.0, but these lines are mainly important for their gradients, and there the values do vary. But you're absolutely right, it deserves a comment.

Your second question: This is a general aspect of PPO and not an 'issue' with our implementation, right? It's true that technically the minibatches are slightly off-policy after the first minibatch update. However, the point is that they are only *slightly* off, so still approximately right and informative. Generally with deep learning, 'approximately right' usually doesn't hurt performance, in fact it sometimes even helps (noise infusion etc). So that shouldn't be a problem, or at least PPO in general is not the right algorithm to use if it is.

**Comment by magnusja** [comment]
Hey @AlexKuhnle, thanks for the explanation. Sure you are right, it should not be too bad. However, in the OpenAI baseline code (where they cache the old policy according to my understanding) and this repo[1] where they cache the policy parameters, it seems to be implemented the way I thought it should be. The PPO paper is also not really clear on that, I think both interpretations are going along with what is described in the paper.

[1] https://github.com/tensorflow/agents

**Comment by magnusja** [comment]
Okay @AlexKuhnle  one follow up question if you do not mind :D

How does clipping the ratio later on make sense if it is always 1?

```python
clipped_prob_ratio = tf.clip_by_value(
                t=prob_ratio,
                clip_value_min=(1.0 / (1.0 + self.likelihood_ratio_clipping)),
                clip_value_max=(1.0 + self.likelihood_ratio_clipping)
            )
return -tf.minimum(x=(prob_ratio * reward), y=(clipped_prob_ratio * reward))
```

**Comment by AlexKuhnle** [comment]
This is an interesting question indeed...^^ I don't have a spontaneous good answer for that, so will have to think about it on the weekend.

**Comment by AlexKuhnle** [comment]
Hey @magnusja,

Finally got around having a closer look at this. Indeed there was a problem the way we de-composed the PPO optimization procedure as a `multi_step(subsampling_step(adam))`, with each step of `multi_step` comparing to the old policy before this step instead of the old policy before the multi-step optimization. The fact that in the first step the ratio equals `1.0`, however, is also mentioned in the paper: *"Note that L^CLIP(θ) = L^CPI(θ) to first order around θ_old (i.e., where r = 1), however, they become different as θ moves away from θ_old."* The latest commit fixes this problem. Thanks very much for bringing this up!

**Comment by Mazecreator** [comment]
@AlexKuhnle Did you mean to push that to memory or master?  Looks like the update went to memory.

**Comment by lefnire** [comment]
@Mazecreator I'm under the impression #master is in something of a freeze right now; all the latest hotness going to #memory

**Comment by Mazecreator** [comment]
Oh, okay, I will pull that one then

**Comment by magnusja** [comment]
Hey @AlexKuhnle,

I quickly skimmed through the code and this looks now much more like the other implementations I found online! Thanks for fixing.

**Comment by AlexKuhnle** [comment]
@Mazecreator, as lefnire says, we will soon merge the new branch anyway, so I just changed it there. It would work similarly on master, but not exactly the same due to some changes.
